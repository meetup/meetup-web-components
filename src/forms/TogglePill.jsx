import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';
import Icon from '../media/Icon';
import ToggleWrapper from '../utils/components/ToggleWrapper';

export const TOGGLE_PILL_CLASS = 'toggleButton';

/**
 * Toggle Pill component
 * @see {@link https://github.com/meetup/sassquatch2/blob/develop/sass/ui-components/_toggle-pill.scss}
 * @see {@link http://meetup.github.io/sassquatch2/ui_components.html#togglePills}
 * @module TogglePill
 */
class TogglePill extends React.PureComponent {
	render() {
		const {
			isActive,
			children,
			className,
			labelClassName,
			topic,
			id,
			large,
			name,
			useRadio,
			value,
			onChange,
			...other
		} = this.props;

		const inputType = useRadio ? 'radio' : 'checkbox';

		const classNames = cx(
			TOGGLE_PILL_CLASS,
			{
				'toggleButton--topic': topic,
				'toggleButton--radio': useRadio
			},
			className
		);

		const labelClassNames = cx(
			'toggleButton-label',
			{
				'toggleButton-label--large': large
			},
			labelClassName
		);

		// ---
		// Topic variant
		const topicClassName = cx(
			'toggleButton-icon',
			{
				'toggleButton-icon--active' : isActive,
				'toggleButton-icon--inactive' : (!isActive)
			}
		);

		const iconShape = isActive ? 'heart' : 'heart-outline';

		const topicChildren = (
			<Icon
				className={topicClassName}
				shape={iconShape}
				size='xxs'
				label='Active Topic Icon'/>
		);
		// ---

		return (
			<ToggleWrapper
				type={inputType}
				isActive={isActive}
				onToggle={onChange}
			>
				{({
					tabIndex,
					isActive,
					toggleActive,
					onKeyUp,
				}) => (
					<div className={classNames}>
						<input
							className='toggleButton-input visibility--a11yHide'
							type={inputType}
							id={id}
							name={name}
							value={value}
							checked={isActive}
							onChange={toggleActive}
							onKeyUp={onKeyUp}
							tabIndex={tabIndex}
							{...other} />
						<label
							className={labelClassNames}
							htmlFor={id}>
							{children}
							{(topic) ? topicChildren : null}
						</label>
					</div>
				)}
			</ToggleWrapper>
		);
	}
}

TogglePill.protoTypes = {
	id: PropTypes.string.isRequired,
	name: PropTypes.string.isRequired,
	value: PropTypes.string.isRequired,
	children: PropTypes.node.isRequired,
	isActive: PropTypes.bool,
	useRadio: PropTypes.bool,
	topic: PropTypes.bool,
	labelClassName: PropTypes.string,
};

TogglePill.defaultProps = {
	isActive: false
};

export default TogglePill;

