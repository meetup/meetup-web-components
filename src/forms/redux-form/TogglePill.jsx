import React from 'react';
import TogglePill from '../TogglePill';

/**
 * @param {Object} props - props passed in from parent and redux-form
 * @description wraps standard TogglePill web component for use with redux-form
 * deconstructs props that redux-forms sets and sets them on TogglePill
 *
 * NOTE: redux-form expects checkboxes value to be true/false and will set them as so
 * https://github.com/erikras/redux-form/issues/2922, therefore using the value here to
 * set isActive
 *
 * @return {Component} TogglePill
 */
const ReduxFormTogglePill = props => {
	// not adding meta error to TogglePill prop for now
	// since TogglePills should function in a group with one error
	const {
		meta, // eslint-disable-line no-unused-vars
		input,
		...other
	} = props;

	return <TogglePill {...input} isActive={input.value === true} {...other} />;
};

ReduxFormTogglePill.displayName = 'ReduxFormTogglePill';

export default ReduxFormTogglePill;
