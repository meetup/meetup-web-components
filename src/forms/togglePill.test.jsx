import React from 'react';
import { mount } from 'enzyme';
import TogglePill from './TogglePill';

describe('TogglePill', () => {
	const id = 'hikingCategory',
		name = 'meetupCategories',
		label = 'Hiking!',
		value = 'hiking';

	let togglePillComponent;
	const onChangeMock = jest.fn();

	beforeEach(() => {
		togglePillComponent = mount(
			<TogglePill onChange={onChangeMock} id={id} name={name} value={value}>
				{label}
			</TogglePill>
		);
	});

	afterEach(() => {
		togglePillComponent = null;
	});

	it('renders a component with expected attributes', () => {
		expect(togglePillComponent).toMatchSnapshot();
	});

	it('executes onChange when clicked', () => {
		expect(onChangeMock).not.toHaveBeenCalled();
		togglePillComponent.simulate('change');
		expect(onChangeMock).toHaveBeenCalled();
	});
});
