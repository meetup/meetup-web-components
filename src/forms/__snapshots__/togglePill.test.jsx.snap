// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TogglePill renders a component with expected attributes 1`] = `
<TogglePill
  id="hikingCategory"
  isActive={false}
  labelClassName="someClass"
  large={true}
  name="meetupCategories"
  onChange={[Function]}
  value="hiking"
>
  <ToggleWrapper
    isActive={false}
    onToggle={[Function]}
    type="checkbox"
  >
    <ConditionalWrap
      condition={false}
      wrap={[Function]}
    >
      <div
        className="toggleButton"
      >
        <input
          checked={false}
          className="toggleButton-input visibility--a11yHide"
          id="hikingCategory"
          name="meetupCategories"
          onChange={[Function]}
          onKeyUp={[Function]}
          tabIndex="0"
          type="checkbox"
          value="hiking"
        />
        <label
          className="toggleButton-label toggleButton-label--large someClass"
          htmlFor="hikingCategory"
        >
          Hiking!
        </label>
      </div>
    </ConditionalWrap>
  </ToggleWrapper>
</TogglePill>
`;
